app/scripts/modules/core/src/securityGroup/SecurityGroups.tsx

```diff
@@ -0,0 +1,153 @@
+import { CreateSecurityGroupButton } from './CreateSecurityGroupButton';
+
+export interface ISecurityGroupsProps {
+  app: Application;
+}
+
+interface IFilterModel {
+  groups: ISecurityGroupGroup[];
+  tags: IFilterTag[];
+}
+
+const Groupings = ({ groups, app }: { groups: ISecurityGroupGroup[]; app: Application }) => (
+  <div>
+    {groups.map(group => (
+      <div key={group.heading} className="rollup">
+        {group.subgroups &&
+          group.subgroups.map(subgroup => (
+            <SecurityGroupPod
+              key={subgroup.heading}
+              grouping={subgroup}
+              application={app}
+              parentHeading={group.heading}
+            />
+          ))}
+      </div>
+    ))}
+    {groups.length === 0 && (
+      <div>
+        <h4 className="text-center">No {FirewallLabels.get('firewalls')} match the filters you've selected.</h4>
+      </div>
+    )}
+  </div>
+);
+
+const Filters = () => {
+  const toggleParam = (event: any): void => {
+    const { checked } = event.target;
+    const name: keyof ISortFilter = event.target.name;
+    SecurityGroupState.filterModel.asFilterModel.sortFilter[name] = checked;
+    SecurityGroupState.filterModel.asFilterModel.applyParamsToUrl();
+  };
+
+  // The way these "hideX" stateParams work is extra confusing and we should never do this inverse-boolean thing again
+  return (
+    <div className="col-lg-8 col-md-10">
+      <div className="form-inline clearfix filters">
+        <div className="form-group">
+          <label className="checkbox"> Show </label>
+          <div className="checkbox">
+            <label>
+              <input
+                type="checkbox"
+                name="showServerGroups"
+                checked={ReactInjector.$stateParams.hideServerGroups === true}
+                onChange={toggleParam}
+              />{' '}
+              Server Groups
+            </label>
+          </div>
+          <div className="checkbox">
+            <label>
+              <input
+                type="checkbox"
+                name="showLoadBalancers"
+                checked={ReactInjector.$stateParams.hideLoadBalancers === true}
+                onChange={toggleParam}
+              />{' '}
+              Load Balancers
+            </label>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export const SecurityGroups = ({ app }: ISecurityGroupsProps) => {
+  const { useState, useEffect } = React;
+  const [filterModel, setFilterModel] = useState<IFilterModel>({ groups: [], tags: [] });
+  const [initialized, setInitialized] = useState(false);
+
+  const groupsUpdated = React.useRef(
+    debounce(() => {
+      setFilterModel({
+        groups: SecurityGroupState.filterModel.asFilterModel.groups,
+        tags: SecurityGroupState.filterModel.asFilterModel.tags,
+      });
+      if (!initialized && app.getDataSource('securityGroups').loaded) {
+        setInitialized(true);
+      }
+    }, 100),
+  ).current;
```

Comment:
If I'm reading this right, what you're trying to do is:

- Pass a stable, debounced reference to a callback into the filter service
- Still have fresh references to the most recent render's variables inside the callback closure (e.g. `initialized`)

If that's accurate, then I don't believe this implementation is actually working as intended — the ref is definitely giving back a stable reference, but the callback is never going to get freshly generated to pick up variables from the outer closure. I suspect that every time this callback runs it's going to call `setInitialized(true)`, because `initialize` inside the callback will always be `false`.

I recommend trying out a different approach that uses [a hook I wrote a couple weeks back called `useLatestCallback`](https://github.com/spinnaker/deck/pull/7659/commits/cd6469bc8ac8c2360b2b704d8f882390283c3bb6) — it wraps around `useRef` and gives you back a stable callback function that will _always_ call whatever the latest function you pass into it is — so if you have a callback that you want regenerated on every render, you pass that fresh callback into the hook and it gives you back the same function reference it gave you last time. Then we can combine that with the native `useCallback` hook which will let you wrap that stabilized callback in a `debounced` call and hold onto it for the lifetime of the component. So in practice:

```tsx
const handleGroupsUpdated = useLatestCallback(() => {
  setFilterModel({
    groups: SecurityGroupState.filterModel.asFilterModel.groups,
    tags: SecurityGroupState.filterModel.asFilterModel.tags,
  });
  if (!initialized && app.getDataSource('securityGroups').loaded) {
    setInitialized(true);
  }
});
// using an empty array means the debounced callback will never get regenerated
const groupsUpdated = useCallback(debounce(handleGroupsUpdated, 100), []);
```

Reply:
It's actually just a weird way (thanks first google search result) to get a debounced function across renders, but it's cleaner to do it with `useCallback`, so I'll update it.

Reply:
I got rid of the debounce altogether.

Reply:
the real debounce was the fun stuff about hooks we learned along the way 🌈